### PR TITLE
Initialization and shutdown statemachine

### PR DESCRIPTION
**Pull request by jdmnd**

### CODE REVIEW

File: src/deepstream.io.js [hunk 1]
```diff
@@ -121,44 +155,23 @@ Deepstream.prototype.isRunning = function () {
...
-
-  let i
-  let initialiser
-
-  this._options.pluginTypes.forEach((pluginType) => {
-    initialiser = new DependencyInitialiser(this._options, pluginType)
-    initialiser.once('ready', this._checkReady.bind(this, pluginType, initialiser.getDependency()))
+  process.nextTick(() => this._state.start())
+
+  process.on('SIGINT', () => {
```
Comment (yasserf) [inline src/deepstream.io.js]: this is tricky.. isn't it done in the bin directory?

Comment (jdmnd) [inline src/deepstream.io.js]: You're right, this made it into the PR by mistake. See the changes to bin/deepstream-start.js

File: conf/config.yml [hunk 2]
```diff
@@ -9,24 +9,41 @@ logLevel: INFO
...
+        # the amount of milliseconds between each ping/heartbeat message
+        pingInterval: 30000
+
+        # Security
+        # amount of time a connection can remain open while not being logged in
+        unauthenticatedClientTimeout: 180000
+        # invalid login attempts before the connection is cut
+        maxAuthAttempts: 3
+        # if true, the logs will contain the cleartext username / password of invalid login attempts
+        logInvalidAuthData: true
```
Comment (yasserf) [inline conf/config.yml]: should we change this to false by default then? ;)

File: src/config/config-initialiser.js [hunk 3]
```diff
@@ -166,14 +167,54 @@ function handlePlugins (config) {
...
-      if(config.pluginTypes.indexOf(key) === -1) {
+      if (config.pluginTypes.indexOf(key) === -1) {
         config.pluginTypes.push(key)
       }
     }
   }
 }
 
 /**
+ *
```
Comment (yasserf) [inline src/config/config-initialiser.js]: missing comment

File: src/config/config-initialiser.js [hunk 4]
```diff
@@ -166,14 +167,54 @@ function handlePlugins (config) {
...
+ */
+function handleConnectionEndpoints (config) {
+  if (!config.connectionEndpoints || Object.keys(config.connectionEndpoints).length === 0) {
+    throw new Error('No connection endpoints configured')
+  }
+  if (Object.keys(config.connectionEndpoints).length > 1) {
+    throw new Error('Currently only one connection endpoint may be configured')
+  }
+  const connectionEndpoints = []
+  for (const connectionType in config.connectionEndpoints) {
```
Comment (yasserf) [inline src/config/config-initialiser.js]: cant we now do something like

const [ connectionType, plugin ] in config.connectionEndpoints?

Don't need to change, just curious

File: src/message/uws-socket-wrapper.js [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    this._logger = logger
+    this.user = null
+    this.authCallBack = null
+    this.authAttempts = 0
+    this.setMaxListeners(0)
+    this.uuid = Math.random()
+    this._handshakeData = handshakeData
+    this._external = external
+
+    this._queuedMessages = []
```
Comment (yasserf) [inline src/message/uws-socket-wrapper.js]: this four properties seem to not be used

File: test.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
+const deepstream = require('./src/deepstream.io')
```
Comment (yasserf) [inline test.js]: probably don't want this in there

File: test/bin/installerSpec.js [hunk 7]
```diff
@@ -3,7 +3,7 @@
 
 const path = require('path')
 const pj = path.join
-const proxyquire = require('proxyquire')
+const proxyquire = require('proxyquire').noPreserveCache()
```
Comment (yasserf) [inline test/bin/installerSpec.js]: so we ended up using this?

**Comment by yasserf** [comment]
Added comments

File: conf/config.yml (hunk 2 above)
Comment (jdmnd) [inline conf/config.yml]: is changing a default not a breaking change?

File: src/config/config-initialiser.js (hunk 4 above)
Comment (jdmnd) [inline src/config/config-initialiser.js]: destructuring isn't node 4 compatible

Comment (jdmnd) [inline src/config/config-initialiser.js]: but possibly could use for...of

Comment (jdmnd) [inline src/config/config-initialiser.js]: but possibly could use for...of

File: test.js (hunk 6 above)
Comment (jdmnd) [inline test.js]: 😔

File: conf/config.yml (hunk 2 above)
Comment (jdmnd) [inline conf/config.yml]: nah

**Pull request merged**